packages/core/crates/core-crypto/src/routing.rs

```diff
@@ -0,0 +1,381 @@
+use crate::derivation::derive_mac_key;
+use crate::errors::CryptoError;
+use crate::errors::CryptoError::TagMismatch;
+use crate::errors::Result;
+use crate::parameters::SECRET_KEY_LENGTH;
+use crate::prg::{PRGParameters, PRG};
+use crate::primitives::{create_tagged_mac, DigestLike, SimpleMac};
+use crate::random::random_fill;
+use crate::routing::ForwardedHeader::{FinalNode, RelayNode};
+use crate::types::PublicKey;
+use crate::utils::xor_inplace;
+use std::ops::Not;
+use subtle::ConstantTimeEq;
+
+const RELAYER_END_PREFIX: u8 = 0xff;
+
+/// Returns the size of the packet header given the information about the number of hops and additional relayer info.
+pub const fn header_length(
+    max_hops: usize,
+    additional_data_relayer_len: usize,
+    additional_data_last_hop_len: usize,
+) -> usize {
+    let per_hop = PublicKey::SIZE_COMPRESSED + SimpleMac::SIZE + additional_data_relayer_len;
+    let last_hop = 1 + additional_data_last_hop_len;
+
+    last_hop + (max_hops - 1) * per_hop
+}
+
+fn generate_filler(
+    max_hops: usize,
+    routing_info_len: usize,
+    routing_info_last_hop_len: usize,
+    secrets: &[&[u8]],
+) -> Box<[u8]> {
+    if secrets.len() < 2 {
+        return vec![].into_boxed_slice();
+    }
+
+    assert!(max_hops >= secrets.len(), "too few hops");
+    assert!(routing_info_len > 0, "invalid routing info length");
+
+    let header_len = routing_info_last_hop_len + (max_hops - 1) * routing_info_len;
+    let padding_len = (max_hops - secrets.len()) * routing_info_len;
+
+    let mut ret = vec![0u8; header_len - padding_len - routing_info_last_hop_len];
```

Comment:
Is the `0u8` initialization necessary? The capacity option will allocate zeroed pages, so zeroing is not necessary, the original version might obfuscate intent, this is clearer:

```suggestion
    let mut ret = Vec::with_capacity(header_len - padding_len - routing_info_last_hop_len);
```

Reply:
The `with_capacity` only creates vector with zeroed pages, but the vector does not have such *length* yet.
As such, the vector would then need to be resized as well. And since `resize` also re-initializes the entries, I didn't see much difference between that and creating an already zero-ed out new vector.